src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,390 @@
+            # Regex '[abc0-9]' (set of characters)
+            negate = value[0][0] == sre.NEGATE
+            if isinstance(pattern, text_type):
+                builder = CharactersBuilder(negate, context.flags)
+            else:
+                builder = BytesBuilder(negate, context.flags)
+
+            for charset_code, charset_value in value:
+                if charset_code == sre.NEGATE:
+                    # Regex '[^...]' (negation)
+                    # handled by builder = CharactersBuilder(...) above
+                    pass
+                elif charset_code == sre.LITERAL:
+                    # Regex '[a]' (single char)
+                    builder.add_char(charset_value)
+                elif charset_code == sre.RANGE:
+                    # Regex '[a-z]' (char range)
+                    low, high = charset_value
+                    for char_code in hrange(low, high + 1):
+                        builder.add_char(char_code)
+                elif charset_code == sre.CATEGORY:
+                    # Regex '[\w]' (char category)
+                    builder.add_category(charset_value)
+                else:  # pragma: no cover
+                    # Currently there are no known code points other than
+                    # handled here. This code is just future proofing
+                    raise AssertionError('Unknown charset code: %s'
+                                         % charset_code)
+            return builder.strategy
+
+        elif code == sre.ANY:
+            # Regex '.' (any char)
+            if isinstance(pattern, text_type):
+                if context.flags & re.DOTALL:
+                    return st.characters()
+                return st.characters(blacklist_characters=u'\n')
+            else:
+                if context.flags & re.DOTALL:
+                    return binary_char
+                return binary_char.filter(lambda c: c != b'\n')
+
+        elif code == sre.AT:
+            # Regexes like '^...', '...$', '\bfoo', '\Bfoo'
+            # An empty string (or newline) will match the token itself, but
+            # we don't and can't check the position (eg '%' at the end)
+            if value == sre.AT_END:
+                newline = u'\n' if isinstance(pattern, text_type) else b'\n'
+                return st.sampled_from([empty, newline])
+            return st.just(empty)
+
+        elif code == sre.SUBPATTERN:
+            # Various groups: '(...)', '(:...)' or '(?P<name>...)'
+            old_flags = context.flags
+            if HAS_SUBPATTERN_FLAGS:  # pragma: no cover
+                # This feature is available only in specific Python versions
+                context.flags = (context.flags | value[1]) & ~value[2]
+
+            strat = _strategy(value[-1], context, pattern)
+
+            context.flags = old_flags
+
+            if value[0]:
+                context.groups[value[0]] = strat
+                strat = st.shared(strat, key=(pattern, value[0]))
```

Comment:
> that's not what happens here

It's not what happens here because you've got the filter for match at the end. I'm pretty sure it is what happens here without that.

shared shares things so that for each key argument a value is drawn at most once. So if you do `shared(x, k)` and `shared(y, k)` then whichever gets drawn first "claims" the key and stuffs its value into that, and the other one will draw that value instead of drawing from the underlying strategy. Think of it as the `setdefault` strategy maybe.

> it's going to be a nastier and more stateful strategy

It's not going to be *much* nastier and more stateful. It basically requires a single top level wrapper somewhere that looks more or less like this:

```
REGEX_GROUP_CACHE_STRATEGY = shared(builds(dict), key="hypothesis.regex.group_cache")

@composite
def invalid_cache_after_draw(draw):
    result = draw(base_strategy)
    cache = draw(REGEX_GROUP_CACHE_STRATEGY)
    cache.clear()
    return result
```

> Would you object to merging this as a working-if-imperfect strategy, and opening an issue to improve the niggly bits later?

Sorry, I would. I'm at this point prepared to accept working-if-imperfect, but I'd like to sort out at least the *known* bugs in it before merging!